server/pulp/server/lazy/streamer.py

```diff
@@ -0,0 +1,212 @@
+        """
+        Modify incoming headers to include the cache timeout value from the Pulp
+        server configuration. This is used by Squid to determine what content
+        can be removed from its cache when its reaper runs.
+
+        :param report: The download report for this request.
+        :type  report: nectar.report.DownloadReport
+        """
+        for header_key, header_value in report.headers.items():
+            self.request.setHeader(header_key, header_value)
+
+        max_age = {'max_age': pulp_config.get('lazy', 'streamer_cache_timeout')}
+        cache_header = 'public, s-maxage=%(max_age)s, max-age=%(max_age)s' % max_age
+        self.request.setHeader('cache-control', cache_header)
+
+    def download_failed(self, report):
+        """
+        Perform cleanup on failed downloads. Specifically, Nectar does not download
+        any content if the server returns a non-200 code. This is problematic because
+        it will return headers to the client which include a Content-Length that is
+        likely not 0, since most servers provide pages detailing the problem that
+        occurred.
+
+        :param report: The download report for this request.
+        :type  report: nectar.report.DownloadReport
+        """
+        error_report = report.error_report.copy()
+        error_report['url'] = report.url
+        logger.info(_('Download of %(url)s failed: HTTP %(response_code)s '
+                      '%(response_msg)s' % error_report))
+
+        # Currently Nectar returns headers with a content-length even
+        # when it doesn't download anything.
+        self.request.setHeader('Connection', 'close')
+        self.request.setHeader('Content-Length', '0')
+        if 'response_code' in report.error_report:
+            self.request.setResponseCode(report.error_report['response_code'])
+
+
+class Streamer(resource.Resource):
+    """
+    Define the web resource that streams content from the upstream repository
+    to the client.
+    """
+    # Ensure self.getChild isn't called as this has no child resources
+    isLeaf = True
+
+    def render_GET(self, request):
+        """
+        Handle GET requests to this web resource.
+
+        This method performs the following tasks to service the request:
+
+            * The requested URL is checked to ensure it has a valid signature
+              from the Pulp server.
+            * The unit specified by the request is looked up in the Pulp unit
+              catalog and the correct downloader is retrieved based on the specific
+              file requested.
+            * The file is downloaded using the Nectar downloader and the content
+              is streamed to the client as it is received.
+
+        :param request: the request to process.
+        :type  request: twisted.web.server.Request
+        """
+        reactor.callInThread(self._download, request)
+        return NOT_DONE_YET
+
+    def _download(self, request):
```

Comment:
This method is doing a lot.  Suggest refactoring into several methods with each doing one specific thing.

Also, we'll want to support matching multiple catalog entries and trying them all.  We can discuss.